File: 2021/fsharp/Aoc2021/Day07.fs
module Day07

open Utils

let findPositionWithMinFuel distanceToFuel positions =
    let positionCount =
        positions
        |> List.countBy (fun position -> position)

    let fuel (position: int) =
        positionCount
        |> List.sumBy (fun (pos, count) ->
            count
            * (System.Math.Abs(position - pos) |> distanceToFuel))

    let min = positionCount |> List.minBy fst |> fst
    let max = positionCount |> List.maxBy fst |> fst

    [ min .. max ]
    |> List.map (fun pos -> pos, pos |> fuel)
    |> List.minBy snd

let findBestPositionWithMinFuel = findPositionWithMinFuel (fun d -> d)

let findBestCrabPositionWithMinFuel =
    findPositionWithMinFuel (fun d -> d * (d + 1) / 2)

let input =
    "07" |> readInputText |> String.parseInts ','

let job1 () =
    input
    |> findBestPositionWithMinFuel
    |> snd
    |> string

let job2 () =
    input
    |> findBestCrabPositionWithMinFuel
    |> snd
    |> string

File: 2021/csharp/Aoc2021/Day07.cs
namespace Aoc2021;

public class Day07 : IDay
{
    public override string Day { get; } = nameof(Day07)[3..];

    public override string Result1()
    {
        return FindBestPostion(InputText.Split(',').Select(ParseInt).ToList())
            .fuel
            .ToString(CultureInfo.InvariantCulture);
    }

    public override string Result2()
    {
        return FindBestCrabPostion(InputText.Split(',').Select(ParseInt).ToList())
            .fuel
            .ToString(CultureInfo.InvariantCulture);
    }

    public static (int position, int fuel) FindBestPostion(IEnumerable<int> positions)
    {
        return FindMinFuelPosition(positions, d => d);
    }

    public static (int position, int fuel) FindBestCrabPostion(IEnumerable<int> positions)
    {
        return FindMinFuelPosition(positions, d => d * (d + 1) / 2);
    }

    public static (int position, int fuel) FindMinFuelPosition(
        IEnumerable<int> positions,
        Func<int, int> distanceToFuel
    )
    {
        List<(int position, int count)> positionCount = positions
            .GroupBy(
                position => position,
                (position, positions) => (position, positions.Count())
            )
            .ToList();

        (int position, int fuel) result = Fuels().MinBy(position => position.fuel);
        return result;

        List<(int position, int fuel)> Fuels()
        {
            int min = positionCount.MinBy(pc => pc.position).position;
            int max = positionCount.MaxBy(pc => pc.position).position;

            return Enumerable.Range(min, max - min + 1)
                .Select(position => (position, Fuel(position)))
                .ToList();
        }

        int Fuel(int position)
        {
            return positionCount
                .Select(item => item.count * distanceToFuel(Math.Abs(item.position - position)))
                .Sum();
        }
    }
}
